Assets/Plugins/iOS/Shopify/PaymentSession.swift

```diff
@@ -0,0 +1,221 @@
+//
+//  PaymentSession.swift
+//  UnityBuySDK
+//
+//  Created by Shopify.
+//  Copyright © 2017 Shopify Inc. All rights reserved.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import UIKit
+import PassKit
+
+@objc enum PaymentStatus: Int {
+    case Failed
+    case Cancelled
+    
+    func toString() -> String {
+        switch self {
+        case .Cancelled:
+            return "Cancelled"
+        case .Failed:
+            return "Failed"
+        }
+    }
+}
+
+@objc protocol PaymentSessionDelegate : class {
+    func paymentSessionDidFinish(session: PaymentSession,with status: PaymentStatus)
+    func paymentSession(_ session: PaymentSession, didSelect shippingMethod: PKShippingMethod)
+    func paymentSession(_ session: PaymentSession, didSelectShippingContact shippingContact: PKContact)
+    func paymentSession(_ session: PaymentSession, didAuthorize payment: PKPayment)
+}
+
+@objc class PaymentSession: NSObject {
+
+    /// Semaphore that blocks the PKAuthorizationController till summary items
+    /// or shipping methods are updated based on the user's input
+    let updateSemaphore = DispatchSemaphore(value: 0)
```

Comment:
Almost. `PaymentSession` will block an asynchronous queue that will own the completion closure. The asynchronous queue is blocked until someone calls `didAuthorize` or `didFinishUpdate`. This doesn't block the  main queue, the user can still interact with the App. Other changes to the shipping address and billing address made by the user during this are put onto a queue by PKPaymentAuthorizationController, so other delegate methods are not called until the original completion closure is finished. With the exception of `paymentAuthorizationViewControllerDidFinish`